/****************************************************************************
**
** Copyright (C) 2013 Atmel Corporation and/or its subsidiary(-ies).
** All rights reserved.
**
** LIMITED License Agreement
**
** IT IS IMPORTANT THAT YOU READ THIS AGREEMENT CAREFULLY AND COMPLETELY.
** This Limited License Agreement (“Agreement”) is a legally binding agreement between,
** on one hand, either your employer (if you are acting on behalf of your employer)
** or you (if you are acting on your own behalf) (“Licensee”), and on the other hand,
** Atmel Corporation (“Atmel”).
**
** By installing or using any of the software downloaded or provided with these
** terms (“Licensed Software”), you are indicating that you are binding Licensee
** to the terms of this Agreement, and that you are duly authorized by Licensee to do so.
** If you are not authorized to bind Licensee to the terms of this Agreement,
** or if Licensee does not agree to be bound by all of the terms of this Agreement,
** do not install or use any such software.
**
**
** 1.	Grant of License.  Subject to the terms and conditions of this Agreement,
** Atmel grants Licensee a non-exclusive, non-transferable, non-sublicensable,
** limited license to use the Licensed Software solely in connection with Atmel products
** from SAM9 and SAMA5 families (“Atmel Product”).
**
** 2.	Restrictions.  Except as expressly set forth in Section 1, Licensee will not,
** and will have no right to, (a) use, copy or reproduce any Licensed Software except,
** (b) modify, create derivative works of, sell, distribute, transfer or disclose any
** Licensed Software. Without limiting the generality of the foregoing, Licensee will not,
** and will have no right to, use any Licensed Software for any semiconductor products
** that are not Atmel Products. Licensee will not remove, obscure or alter any trademark,
** copyright or other proprietary rights or ownership notices of Atmel that appear
** in any Licensed Software.
**
** 3.  	Title.  As between the parties, Atmel retains full rights, title, and ownership
** including all patents, copyrights, trade secrets, trade names, trademarks, and other
** intellectual property rights in and to the Licensed Software.
**
** 4.  	No Other Rights.  Except as expressly stated herein, this Agreement does not grant
** Licensee any rights to patents, copyrights, trade secrets, trade names, trademarks
** (whether registered or unregistered), or any other rights, franchises, or licenses
** in respect of the Licensed Software.
**
** 5.	Disclaimer of Warranty and No Support.  ALL LICENSED SOFTWARE IS PROVIDED “AS IS”,
** “WITH ALL FAULTS”, AND WITH NO WARRANTY WHATSOEVER.  ATMEL EXPRESSLY DISCLAIMS ALL
** WARRANTIES OF ANY KIND (WHEHTER EXPRESS, IMPLIED, STATUTORY OR OTHERWISE) IN CONNECTION
** WITH THE LICENSED SOFTWARE OR ANY OTHER ASPECT OF THIS AGREEMENT, INCLUDING, WITHOUT
** LIMITATION ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE,
** TITLE OR NON-INFRINGEMENT OF THIRD PARTY RIGHTS, AND ANY WARRANTIES THAT MAY ARISE
** FROM COURSE OF DEALING, COURSE OF PERFORMANCE OR USAGE OF TRADE.  ATMEL WILL HAVE
** NO OBLIGATION UNDER THIS AGREEMENT TO CORRECT ANY BUGS, DEFECTS OR ERRORS IN
** THE LICENSED SOFTWARE, PROVIDE ANY UPDATES, UPGRADES OR NEW RELEASES OF THE LICENSED SOFTWARE,
** OR OTHERWISE PROVIDE ANY SUPPORT OR MAINTENANCE FOR THE LICENSED SOFTWARE.
**
** 6.  	Export.  Licensee will comply with all applicable laws and regulations of all
** relevant jurisdictions in connection with its activities related to the Licensed Software.
** Without limitation of the foregoing, Licensee acknowledges that certain laws
** and regulations of the United States and other jurisdictions may pertain to the export
** and re-export of the Licensed Software, and Licensee will not export or re-export
** any Licensed Software in any form without the appropriate governmental approvals,
** or otherwise in violation of any such laws or regulations.
**
** 7.	Termination.  The license will automatically terminate if Licensee fails to comply
** with any of the terms and conditions of the license including, without limitation
** the confidentiality obligation or the obligations set forth in Sections 1, 2, 6, 7, 10
** and 11 herein. Upon termination for any reason, Licensee will immediately destroy
** or return to Atmel all whole or partial copies of the Licensed Software.
**
** 8.	High Risk Activities. LICENSEE ACKNOWLEDGES AND AGREES THAT THE LICENSED SOFTWARE
** IS NOT DESIGNED OR APPROVED FOR, AND WILL NOT BE INCORPORATED (WITHOUT THE EXPRESS WRITTEN
** APPROVAL OF AN OFFICER OF ATMEL) INTO, ANY PRODUCTS THAT ARE USED OR DESIGNED TO BE USED
** IN CONNECTION WITH ANY ACTIVITIES WHERE THE FAILURE OF SUCH PRODUCTS COULD REASONABLY
** BE EXPECTED TO RESULT IN DEATH, BODILY INJURY, OR SEVERE PHYSICAL OR ENVIRONMENTAL DAMAGE
** ("HIGH RISK ACTIVITIES").  IN NO EVENT WILL ATMEL HAVE ANY LIABILITY TO LICENSEE
** OR ANY THIRD PARTY ARISING OUT OF OR RELATED TO ANY USE OF LICENSED SOFTWARE IN CONNECTION
** WITH HIGH RISK ACTIVITIES, AND ATMEL HEREBY DISCLAIMS ANY EXPRESS OR IMPLIED WARRANTY OF
** FITNESS FOR ANY HIGH RISK ACTIVITIES.
**
** 9.  	LIMITATION OF LIABILITY.  TO THE EXTENT PERMITTED BY APPLICABLE LAW, IN NO EVENT WILL
** ATMEL BE LIABLE TO LICENSEE OR ANY THIRD PARTY (WHETHER SUCH LIABILITY IS BASED ON CONTRACT,
** NEGLIGENCE, STRICT LIABILITY, OTHER TORT THEORY, CONTRIBUTION, BREACH OF WARRANTY,
** OR OTHER LEGAL OR EQUITABLE THEORY) FOR ANY SPECIAL, INDIRECT, INCIDENTAL, EXEMPLARY,
** PUNITIVE OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES FOR LOSS OF PROFITS,
** LOSS OR INTERRUPTION OF BUSINESS, OR LOSS OF DATA, ARISING OUT OF OR RELATING TO
** THE LICENSED SOFTWARE OR ANY OTHER ASPECT OF THIS AGREEMENT, EVEN IF ATMEL has been advised
** of or should have known of the possibility of such damages.
**
** IN NO EVENT WILL ATMEL’S TOTAL LIABILITY ARISING OUT OF OR RELATED TO THE LICENSED SOFTWARE
** OR ANY OTHER ASPECT OF THIS AGREEMENT (WHETHER UNDER CONTRACT, NEGLIGENCE, STRICT LIABILITY,
** CONTRIBUTION, BREACH OF WARRANTY, OR OTHER LEGAL OR EQUITABLE THEORY) EXCEED THE GREATER OF
** (A) THE AGGREGATE OF ALL LICENSE FEES PAID BY LICENSEE TO ATMEL FOR THE LICENSED SOFTWARE, OR
** (B) ONE THOUSAND DOLLARS ($1,000). WITHOUT LIMITING THE FOREGOING,
** AND Notwithstanding any provision herein to the contrary, ATMEL will not be liable for
** ANY costs of PROCURING SUBSTITUTE GOODS, SERVICES OR TECHNOLOGY under any circumstances.
**
** Each Party acknowledges that THE OTHER PARTY has entered into this agreement in reliance
** on the limitations of liability, DISCLAIMERS OF WARRANTIES, EXCLUSION OF DAMAGES
** AND EXCLUSIVE REMEDIES contained in this AGREEMENT, AND THAT EACH OF THE FOREGOING PROVISIONS FORMS
** AN ESSENTIAL AND fundamental part of the basis of the bargain BETWEEN THE PARTIES,
** WITHOUT WHICH SUCH the other Party would not have entered into this Agreement.
** EACH PARTY AGREES THAT SUCH PROVISIONS WILL SURVIVE AND APPLY NOTWITHSTANDING ANY
** FAILURE OF ESSENTIAL PURPOSE OF ANY LIMITED REMEDY OR LIMITATION OF LIABILITY.
**
** 10.  	Governing Law; Dispute Resolution. This Agreement is to be construed in accordance
** with and governed by the internal laws of the State of California (as permitted by Section 1646.5
** of the California Civil Code or any similar successor provision), without giving effect
** to any choice of law rule that would cause the application of the laws of any jurisdiction
** other than the internal laws of the State of California to the rights and duties of the parties.
** This Agreement will not be governed by the U.N. Convention on the Sale of Goods,
** the application of which is expressly excluded. Except for actions for injunctive
** or other equitable relief, which may be brought in any court of competent jurisdiction,
** all disputes arising out of or related to this Agreement will be subject to the exclusive
** jurisdiction of the California state courts in Santa Clara County, California,
** or if there is exclusive federal jurisdiction, the United States District Court for
** the Northern District of California, and the Parties hereby consent to, and agree to submit to,
** the personal and exclusive jurisdiction and venue of such courts.
**
** 11. 	General. Licensee will not, and will have no right to, assign, delegate or otherwise
** transfer (whether voluntarily, by operation of law or otherwise) this Agreement or any of
** its rights or obligations hereunder to any third party without the prior written consent
** of Atmel, and any purported assignment, delegation or other transfer without such consent
** will have no force or effect. Subject to the foregoing, this Agreement will be binding upon
** and will inure to the benefit of the parties and their respective successors and permitted assigns.
** No failure of either party to enforce any right under this Agreement will be deemed a waiver
** of such right or any other right under this Agreement. Any waiver by a party of a breach
** of any provision of this Agreement by the other party hereunder will not be deemed to be
** a waiver of any subsequent breach of such provision or a waiver of any breach of any other
** provision of this Agreement.
** This Agreement may not be superseded, modified, or amended except in a writing signed by an officer
** of each party. If any provision of this Agreement is determined to be invalid, illegal
** or otherwise unenforceable, such provision will be enforced to the extent possible consistent
** with the intent of the parties, and the remaining provisions of this Agreement will remain in
** full force and effect. This Agreement will be fairly interpreted in accordance with its terms
** and without any strict construction against either party because it was drafted by such party
** or for any other reason. This Agreement will constitute the entire agreement between the parties
** relating to the subject matter hereof, and expressly supersedes and replaces all prior
** and contemporaneous agreements, proposals, quotations, negotiations and communications, written
** or oral, between the parties relating to such subject matter.
**
**
** Software Author: Timesys Corporation (www.timesys.com)
**
** Summary: Qt UI handling class for the Light controls screen. Please consult lightingscreen.ui file for
**          object definitions
**
****************************************************************************/

#include "lightingscreen.h"
#include "ui_lightingscreen.h"
#include "zoneselector.h"
#include "SlidingStackedWidget.h"

/***
 ** Default constructor
 **/
LightingScreen::LightingScreen(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::LightingScreen)
{
    ui->setupUi(this);

    //initial state is ON
    ui->labelROOM1->setStyleSheet("QLabel { color : #f79919 }");
    ui->labelROOM2->setStyleSheet("QLabel { color : #f79919 }");
    ui->labelROOM3->setStyleSheet("QLabel { color : #f79919 }");

    QObject::connect(ui->lightControlROOM1, SIGNAL(switchOnOff(bool)), this, SLOT(switchLight1(bool)));
    QObject::connect(ui->lightControlROOM2, SIGNAL(switchOnOff(bool)), this, SLOT(switchLight2(bool)));
    QObject::connect(ui->lightControlROOM3, SIGNAL(switchOnOff(bool)), this, SLOT(switchLight3(bool)));

    mZoneSelector = new ZoneSelector(this);
    mZoneSelector->setBackgroundColor( QColor("#151A20")); //color was hardcoded in customtoolbutton.cpp
    mZoneSelector->setLineColor( QColor("#f79919")); //similar
    QFont font;
    font.setFamily(QString::fromUtf8("Sans Serif"));
    connect( mZoneSelector, SIGNAL(zoneSelected(QString)), this, SLOT(onZoneChanged(QString)));

    if (isSmallResolution) {
        font.setPointSize(12);
        mZoneSelector->setButtonFont( font);
        mZoneSelector->setTitlePixmap( QPixmap(":/resources/smalltopnav-all/title-lighting.png"));
        mZoneSelector->setCanHide( true);
        ui->widgetGraphicsBox->stackUnder( mZoneSelector);

        // float wSFactor = 480. / 800.;
        // float hSFactor = 272. / 480.;

        float fontScale = .9;

        QFont f = ui->labelROOM1->font();
        f.setPointSize(f.pointSize() * fontScale);
        ui->labelROOM1->setFont(f);

        f = ui->labelROOM2->font();
        f.setPointSize(f.pointSize() * fontScale);
        ui->labelROOM2->setFont(f);

        f = ui->labelROOM3->font();
        f.setPointSize(f.pointSize() * fontScale);
        ui->labelROOM3->setFont(f);

        // Relayout
        SlidingStackedWidget *stackWidget = new SlidingStackedWidget(this);
        stackWidget->setObjectName("stackedWidget");
        stackWidget->setWrap(true);

        QPushButton *buttonBack = new QPushButton(this);

        // Reuse image, just rotate it
        QImage img = QImage(":/resources/smallcamera/up.png");
        QMatrix matrix;
        matrix.rotate(-90.);
        img = img.transformed(matrix);
        buttonBack->setIcon(QIcon(QPixmap::fromImage(img)));
        buttonBack->setFlat(true);
        buttonBack->setFocusPolicy(Qt::NoFocus);

        connect (buttonBack, SIGNAL(clicked()), stackWidget, SLOT(slideInPrev()));
        ui->gridLayout->addWidget(buttonBack);

        QIcon icon;
        icon.addFile(QString::fromUtf8(""), QSize(), QIcon::Normal, QIcon::Off);

        // This is more of a workaround then actuall solution because elements on Ui are hardcoded in count
        // instead of being created base on names count

        // Stack widget page 1

        int topContentMargin = 00;

        QWidget *widget = new QWidget(this);
        QVBoxLayout *layout = new QVBoxLayout(widget);
        layout->setContentsMargins(0, topContentMargin, 0, 0);
        ui->gridLayout->removeItem(ui->verticalLayoutLight2);

        layout->addWidget(ui->labelROOM1);
        layout->addWidget(ui->lightControlROOM1);

        // Clean up

        for (int i = ui->verticalLayoutLight2->children().count()-1; i > -1; --i) {
            delete ui->verticalLayoutLight2->children().at(i);
        }

        delete ui->verticalLayoutLight2;

        widget->setLayout(layout);
        stackWidget->addWidget(widget);

        // Stack widget page 2

        widget = new QWidget();
        layout = new QVBoxLayout(widget);
        layout->setContentsMargins(0, topContentMargin, 0, 0);
        ui->gridLayout->removeItem(ui->verticalLayout_5);

        layout->addWidget(ui->labelROOM3);
        layout->addWidget(ui->lightControlROOM3);

        // Clean up

        for (int i = ui->verticalLayout_5->children().count()-1; i > -1; --i) {
            delete ui->verticalLayout_5->children().at(i);
        }

        delete ui->verticalLayout_5;

        widget->setLayout(layout);
        stackWidget->addWidget(widget);

        // Stack widget page 3

        widget = new QWidget();
        layout = new QVBoxLayout(widget);
        layout->setContentsMargins(0, topContentMargin, 0, 0);
        ui->gridLayout->removeItem(ui->verticalLayerLight1);

        layout->addWidget(ui->labelROOM2);
        layout->addWidget(ui->lightControlROOM2);

        // Clean up

        for (int i = ui->verticalLayerLight1->children().count()-1; i > -1; --i) {
            delete ui->verticalLayerLight1->children().at(i);
        }

        delete ui->verticalLayerLight1;

        widget->setLayout(layout);
        stackWidget->addWidget(widget);

        ui->gridLayout->addWidget(stackWidget);

        QPushButton *buttonNext = new QPushButton(this);

        // Reuse image, just rotate it
        img = QImage(":/resources/smallcamera/up.png");
        matrix.rotate(180.);
        img = img.transformed(matrix);
        buttonNext->setIcon(QIcon(QPixmap::fromImage(img)));
        buttonNext->setFlat(true);
        buttonNext->setFocusPolicy(Qt::NoFocus);

        connect (buttonNext, SIGNAL(clicked()), stackWidget, SLOT(slideInNext()));
        ui->gridLayout->addWidget(buttonNext);

        ui->help_maincontrols_Layout->removeItem(ui->verticalSpacer);
        delete ui->verticalSpacer;
        delete ui->help_maincontrols_Layout;
    } else {
        font.setPointSize(16);
        mZoneSelector->setButtonFont( font);
        mZoneSelector->setTitlePixmap( QPixmap(":/resources/topnav-all/title-lighting.png"));
        ui->verticalLayout_2->insertWidget( 0, mZoneSelector);
    }
}

/***
 ** Default destructor
 **/
LightingScreen::~LightingScreen()
{
    delete ui;
}

/***
 ** Sets Zone Names based on information from the Settings panel
 ** Uses defaults at startup
 **/
void LightingScreen::setLighingRoomNames(QList<QStringList> &my2DArray)
{
    roomNames.clear();
    mZoneList.clear();
    for(int i=0; i<my2DArray.size(); i++){
        QStringList strList = (my2DArray[i]);
        roomNames << strList;
        mZoneList << strList[0];
    }
    mZoneSelector->setZoneList( mZoneList);
    switchNames(0);
}

void LightingScreen:: resizeEvent(QResizeEvent*)
{
    mZoneSelector->setWidth( width());
    qDebug() << "lighting zs" << mZoneSelector->geometry();
}

/***
 ** Sets Room Names based on information from the Settings panel
 ** Uses defaults at startup
 **/
void LightingScreen::switchNames(int index)
{
    QStringList* defaultList = (QStringList*)(&(roomNames[index]));

    if(defaultList->size() > 1)
        ui->labelROOM1->setText(defaultList->at(1));
    else
        ui->labelROOM1->setText(QString("Undefined Name"));

    if(defaultList->size() > 2)
        ui->labelROOM2->setText(defaultList->at(2));
    else
        ui->labelROOM2->setText(QString("Undefined Name"));

    if(defaultList->size() > 3)
        ui->labelROOM3->setText(defaultList->at(3));
    else
        ui->labelROOM3->setText(QString("Undefined Name"));
}

/***
 ** Handler for the Room On/Off controls (center of each light control knob)
 ** Color changes - orange:ON, white:OFF
 **/
void LightingScreen::switchLight1(bool status)
{
    if(status)
        ui->labelROOM1->setStyleSheet("QLabel { color : #f79919 }");
    else
        ui->labelROOM1->setStyleSheet("QLabel { color : white; }");

    //TODO: Add code that handles external device (ON/OFF) e.g. via GPIO lines
}

/***
 ** Handler for the Room On/Off controls (center of each light control knob)
 ** Color changes - orange:ON, white:OFF
 **/
void LightingScreen::switchLight2(bool status)
{
    if(status)
        ui->labelROOM2->setStyleSheet("QLabel { color : #f79919 }");
    else
        ui->labelROOM2->setStyleSheet("QLabel { color : white; }");

    //TODO: Add code that handles external device (ON/OFF) e.g. via GPIO lines

}

/***
 ** Handler for the Room On/Off controls (center of each light control knob)
 ** Color changes - orange:ON, white:OFF
 **/
void LightingScreen::switchLight3(bool status)
{
    if(status)
        ui->labelROOM3->setStyleSheet("QLabel { color : #f79919 }");
    else
        ui->labelROOM3->setStyleSheet("QLabel { color : white; }");

    //TODO: Add code that handles external device (ON/OFF) e.g. via GPIO lines

}

void LightingScreen::onZoneChanged(QString zone)
{
    switchNames( mZoneList.indexOf( zone));
}

//TODO: Additional methods can be defined to handle light intensity (dimmer)
